In [2]:
import pandas as pd
import numpy as np

In [3]:
df_general = pd.read_csv('online_retail.csv')

In [4]:
df_general.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [5]:
# Agregar columna de 'totalPrice'
df_general['totalPrice'] = df_general['Quantity'] * df_general['UnitPrice']

In [6]:
# Sustituir los valores nulos en 'CustomerID' por el valor de 'InvoiceNo' en la misma fila + 'A'
df_general['CustomerID'] = df_general['CustomerID'].fillna(df_general['InvoiceNo'] + 'A')

In [7]:
df_general.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   541909 non-null  object 
 7   Country      541909 non-null  object 
 8   totalPrice   541909 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 37.2+ MB


In [8]:
# Agrupa los registros por 'CostumerID' y muestra los 20 clientes que mas gastaron
df_general = df_general.groupby('CustomerID').agg({'totalPrice': 'sum'}).reset_index()
df_general = df_general.sort_values(by='totalPrice', ascending=False).head(20)

In [9]:
df_general.head()

CustomerID  totalPrice
1703    14646.0   279489.02
4233    18102.0   256438.49
3758    17450.0   187482.17
1895    14911.0   132572.62
55      12415.0   123725.45

In [10]:
df = pd.read_csv('online_retail.csv')

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


### Resumen 
1. Se analizaron los nulos de CustomerID para entender el comportamientos de estos datos. 
2. Hay una imposibilidad de determinar los datos devueltos con su espejo positivo, dado que al no tener ID no se determina cual es..
3. Los datos de invoceNo es el mismo independientemente de cuantas veces se repite por los diferentes productos
4. InoviceNo = Misma compra = Mismo ID? (Se supone que el mismo InvoiceNo le pertence al mismo ID)
5. Se procede a sustituir el CustomerID nulos con el valor de su InvoiceNo y se le añade una 'A' para poder identificarlo en el dataframe general. 
6. Se pierde la posibilidad de darle seguimiento a los producto, dado que los invoiceNo en este caso son diferente, pero esto representa el 0,2% de datos total. 
7. Se procede a analizar los datos ya completados. 

In [12]:
#Dime cuales son los CustomerID que más han comprado
df_customer_count = df['CustomerID'].value_counts().reset_index()
df_customer_count.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4372 entries, 0 to 4371
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   CustomerID  4372 non-null   float64
 1   count       4372 non-null   int64  
dtypes: float64(1), int64(1)
memory usage: 68.4 KB


In [15]:
#Cuentame cuantas Quantity  han comprado cada CustomerID
df_quantity_count = df.groupby('CustomerID')['Quantity'].sum().reset_index()
df_quantity_count = df_quantity_count.sort_values(by='Quantity', ascending=False).head(20)  

In [16]:
df_quantity_count.head()

CustomerID  Quantity
1703     14646.0    196719
55       12415.0     77242
1895     14911.0     77180
3758     17450.0     69029
4233     18102.0     64122

In [17]:
#Cuantas Quantity hay en total
total_quantity = df['Quantity'].sum()
print(f'Total Quantity: {total_quantity}')

Total Quantity: 5176450


## Resumen parte2
1. Existe 5.176.450 compras realizadas en total. 
2. El cliente cuyo CustomerID '14646' realizo la mayor cantidad de compras de productos con un total de 196.719 
3. El cliente 